rclrs/src/node.rs

```diff
@@ -71,6 +71,9 @@ pub struct Node {
     pub(crate) guard_conditions_mtx: Mutex<Vec<Weak<GuardCondition>>>,
     pub(crate) services_mtx: Mutex<Vec<Weak<dyn ServiceBase>>>,
     pub(crate) subscriptions_mtx: Mutex<Vec<Weak<dyn SubscriptionBase>>>,
+    _clock: Arc<Mutex<Clock>>,
```

Comment:
Since the inner `_rcl_clock` is an `Arc<Mutex<rcl_clock_t>>` is it really necessary for this `_clock` to also be wrapped in an `Arc<Mutex<Clock>>`? Its only other field is a `ClockType` which is a simple enum so it's definitely `send` and `sync`.

I would recommend making this field a simple `Clock` and changing `get_clock`  to return `&Clock`.

Reply:
I am quite new at `Arc`, `Mutex` and Rust multithreading so I'll admit not being 100% sure of what I am doing.

The same `Clock` instance has to be shared between the node (so users can do a `node.get_clock().now()`) and the `TimeSource` (so it can be driven by a `/clock` topic) so shared ownership is a necessity and `Arc` makes it a lot easier than having to deal with lifetimes.

However, what you say is true and since the inner `_rcl_clock` is wrapped and thread safe, there is technically no need to have this as an `Arc<Mutex<Clock>>`.
In https://github.com/ros2-rust/ros2_rust/pull/325/commits/fde0f35a85d2315b87323cec127d2b1c2c9c41de I changed it from `Arc<Mutex<Clock>>` to `Arc<Clock>`, removing the cost of double locking and making the user face API a lot cleaner. I'm not 100% sure how to remove the need for `Arc` because of the sharing I mentioned above but hopefully this is already a great win, thanks for catching it!
In the process of removing mutexes I also realized that we can just use atomic booleans instead of having a `Mutex<bool>` and addressed that as well https://github.com/ros2-rust/ros2_rust/pull/325/commits/b808cc4bdd06ef6a76308a3aa07e996bbfd71847, it was a fun investigation

Reply:
Done in https://github.com/ros2-rust/ros2_rust/pull/325/commits/5df379fd4c4c162664e2cd8a6ff09e0056f1f1b4 (and cleaned up a bit in following commits).
I tried to use the `RwLock` but encountered poor results. In principle it should help with a many-readers pattern but in practice even the `now` function requires mutable access because the [C binding uses a mutable raw pointer](https://github.com/luca-della-vedova/ros2_rust/blob/524475d5867586f0b8d7c22f677665c0c30cbbb7/rclrs/src/clock.rs#L100) so I couldn't get it to compile.

I left a `new` function to avoid overcomplicating the Node's builder [code](https://github.com/luca-della-vedova/ros2_rust/blob/524475d5867586f0b8d7c22f677665c0c30cbbb7/rclrs/src/node/builder.rs#L274), but can take it out if you think it clutters the API too much.

The main difference with the suggested implementation is in the fact that `ClockSource` has both a function to set the time and one to enable / disable the time override. I believe the only ways to get rid of the enable / disable logic would be to encode it in the `ClockSource` existence (i.e. automatically enable the override when a `ClockSource` is constructed, disable when it is destructed) or diverge a bit from the other client library and have `RosTime` only encode "Time as dictated by the `/clock` topic", and have it fallback to `SystemTime` if `use_sim_time=false`.